Search: add `repo:has.topic()` predicate

camdencheek opened the pull request:
This adds first-class support for searching repos by GitHub topic.

I added automated ingestion of GitHub topics in https://github.com/sourcegraph/sourcegraph/pull/47635. This PR add support for searching repos using this information.

Overview of this PR: 
- Add a GIN index to the metadata stored in Postgres. This enables fast queries on topics even at the scale of sourcegraph.com (I did some manual perf testing to ensure this index can be used).
- Add the `repo:has.topic()` predicate, which exposes this via the search language
- Add hovers, completions, and documentation

This is an alternative to the planned implementation where we ingest topics as KVPs then query them with `repo:has.tag(topic)`. I decided to switch course for a few reasons:
1) It's been brought up that `repo:has.tag()` is confusing because it suggests git tags. `repo:has.topic()` has a meaning that is more obvious.
2) Keeping the `repo_kvp` table consistent with the `repo` table is surprisingly difficult, and there were a lot of edge cases like conflicting keys that would be difficult to solve in a maintainable way. 
3) It keeps repo syncing and search unentangled from each other and greatly simplifies the implementation complexity.

## Test plan

I added a handful of unit tests, database tests, and integration tests to exercise the new behavior. I manually tested that the added index works well on sourcegraph.com.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.33 kb) | 0.00% (+0.64 kb) | 0.00% (+0.31 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/cc/repo-has-topic/compare-report.html;tab=live_object) for a full comparison between the commits e5d9b6a and 09ef967 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

camdencheek commented:
I will add integration tests for this before I merge, but I wanted to get it out for review early because I will be out tomorrow, which starts to push this close to the branch cut

jtibshirani commented on doc/code_search/reference/language.md:
```diff
@@ -616,6 +616,22 @@ Search only inside repositories that contain file content matching the regular e
+
+Search only inside repositories that have the given GitHub topic.
+
+**Example:** [`repo:has.topic(go)` ↗](https://sourcegraph.com/search?q=context:global+repo:has.topic%28go%29&patternType=standard&sm=0&groupBy=repo)
```
Small comment: in reference docs, it'd be nice to use a topic that's not a language to show the power of the feature. Otherwise users might be thinking "why don't I just use `lang: go`" ?

camdencheek replied:
Great idea, thanks!

jtibshirani reviewed (commented):
This  is looking good to me, but I still need to review the database logic. To check -- do you want to ship this with the upcoming release (so we should be speedy with reviews!)

One overall question: am I understanding right that we'll now have both `has.tag` and `has.topic` predicates? This feels a little confusing, will we eventually remove `has.tag`?

camdencheek commented:
>  do you want to ship this with the upcoming release

Yes! We have quite a few customers asking for this, so I'd really like to not make them wait until June 🙂 

> we'll now have both has.tag and has.topic predicates?

Yes, that is correct. `repo:has.tag()` is still for arbitrary user-added key-value-pairs (it's just a key-value pair with a null value). `repo:has.topic()` is specifically for topics as defined by the codehost. We do not support adding topics in Sourcegraph.

I agree this is a little confusing. I would love to do a little bit of consolidation of our query language, and the kvp predicates could use some careful design.

camdencheek commented:
Fixed backend integration tests. The problem is I added topics to the repos in `github.com/sgtest`, but we actually use `ghe.sgdev.org/sgtest` for integration tests 🤦

camdencheek commented:
>  I still need to review the database logic

I added a few comments to the code in an attempt to help clarify what's going on there

jtibshirani commented on doc/code_search/reference/queries.md:
```diff
@@ -114,6 +114,7 @@ The following keywords can be used on all searches (using [RE2 syntax](https://g
 | **fork:yes, fork:only** | Include results from repository forks or filter results to only repository forks. Results in repository forks are excluded by default. | [`fork:yes repo:sourcegraph`](https://sourcegraph.com/search?q=fork:yes+repo:sourcegraph) |
 | **archived:yes, archived:only** | The yes option, includes archived repositories. The only option, filters results to only archived repositories. Results in archived repositories are excluded by default. | [`repo:sourcegraph/ archived:only`](https://sourcegraph.com/search?q=repo:%5Egithub.com/sourcegraph/+archived:only) |
 | **repo:has.path(...)** | Conditionally search inside repositories only if they contain a file path matching the regular expression. See [built-in predicates](language.md#built-in-repo-predicate) for more. | [`repo:has.path(\.py) file:Dockerfile pip`](https://sourcegraph.com/search?q=context:global+repo:has.path%28%5C.py%29+file:Dockerfile+pip&patternType=lucky) |
+| **repo:has.topic(...)** | Search only in repos repositories if they have the given GitHub tag. See [built-in predicates](language.md#built-in-repo-predicate) for more. | [`repo:has.topic(go) file:Dockerfile`](https://sourcegraph.com/search?q=context:global+repo:has.topic%28go%29+file:Dockerfile&patternType=standard&sm=1) |
```
Should we update this to `repo:has.topic(code-search)` too?

camdencheek replied:
Ope, yep! Good catch, thanks.

jtibshirani commented on internal/database/repos.go:
```diff
@@ -1167,6 +1177,29 @@ func (s *repoStore) listSQL(ctx context.Context, tr *trace.Trace, opt ReposListO
+			// the repo's metadata. This is designed to work with the
+			// idx_repo_github_topics index.
+			//
+			// We use the unusual `jsonb_build_array` and `jsonb_build_object`
```
I was wondering if we ever use prepared statements to optimize common queries and prevent SQL injections?

camdencheek replied:
There are two parts to this answer:

For query optimization, our Postgres driver [automatically prepares and caches prepared statements](https://pkg.go.dev/github.com/jackc/pgx/v5#ConnConfig). I haven't actually looked into how well that cache is utilized -- this would be worth looking into. For something like listing repos though, explicitly preparing all variations of the query would likely become untenable given the number of different options we support (all of which change the shape of the query, making the prepared statement unusable).

For SQL injections, `sqlf.Sprintf()` does actually protect against this. It is not the same as `fmt.Sprintf()`. It maintains the separation of query text and query variables which we then [pass separately to Postgres](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3c44710cef763d94d53ea45be97b3ee7705baf59/-/blob/internal/database/basestore/store.go?L86-86), which handles the rest. 

My comment here was just about not being able to construct a JSONB object safely with the string syntax. `SELECT jsonb_build_object('Name', $1::text);` is valid, `SELECT '{"Name": "$1"}'::jsonb;` is not.

jtibshirani commented on migrations/frontend/1678175532_index_github_topics/down.sql:
```diff
@@ -0,0 +1,4 @@
+-- Undo the changes made in the up migration
+
+DROP INDEX IF EXISTS idx_repo_github_topics;
```
In terms of testing -- am I understanding right that CI automatically runs basic checks against migrations (like for idempotency, running them before starting unit tests, etc.)

camdencheek replied:
Yes! We test that all migrations run, they are idempotent, and they can be re-applied after migrating down. All unit tests run against a fully migrated database. I don't know of any good ways to test "does this index apply to my query" other than doing a manual test on a realistic instance and ensuring the `EXPLAIN` output looks right -- would be interested in solutions here 🙂

jtibshirani reviewed (approved):
This looks good to me. I gave my best shot at a review, but I'm still getting up to speed on our SQL standards and approach to database migrations.

I left a couple questions but didn't see anything blocking.

jtibshirani commented on doc/code_search/reference/language.md:
```diff
@@ -616,6 +616,22 @@ Search only inside repositories that contain file content matching the regular e
 
 _Note:_ `repo:contains.content(...)` is an alias for `repo:has.content(...)` and behaves identically.
 
+### Repo has topic
```
It'd also be nice to update these high level docs on repository metadata with the new option: https://docs.sourcegraph.com/admin/repo/metadata

camdencheek replied:
I updated this page, but I did not add any notes about `repo:has.topic()` because this page is specifically about _custom_ metadata. I clarified that in the title and replaced the topics example with one that is a better demonstration of the "custom" part of this feature.
